ARCs/arc-0032.md

```diff
@@ -0,0 +1,233 @@
+  desc: string;
+  max_keys: number;
+}
+
+```
+
+### Type Specification
+
+TODO: rename `userdefinedtype` to something else, user is not a great name for it. custom type? 
+
+
+Each user defined type is specified as either an array of `StructElement`s or an `ABIType`. 
+
+If an array of `StructElement`s is provided, the `UserDefinedType` is treated as a Tuple with named fields.  The ABI encoding is exactly as if an ABI Tuple type defined the same element types in the same order.  It is important to encode the struct elements as an array since it preserves the order of fields which is critical to encoding/decoding the data properly.
+
+If a single ABIType is provided, the `UserDefinedType` is treated as a Type Alias and can be encoded or decoded as the `ABIType` directly.
+
+```ts
+// Type aliases for readability
+type FieldName = string;
+type ABIType = string;
+
+// Each field in the struct contains a name and ABI type
+type StructElement = [FieldName, ABIType];
+
+// Type aliases for readability
+type UserDefinedType = StructElement[] | ABIType;
+type UserDefinedTypeName = string;
+
+// A map of type name to either a list of struct fields or a single ABI type 
+type TypeSpec = Record<UserDefinedTypeName, UserDefinedType>
+```
+
+For example a `UserDefinedType` that should provide an array of `StructElement`s
+
+Given the PyTeal:
+```py
+from pyteal import abi
+
+class Thing(abi.NamedTuple):
+  addr: abi.Field[abi.address]
+  balance: abi.Field[abi.Uint64]
+```
+the equivalent ABI type is `(address,uint64)` and an element in the TypeSpec is:
+
+```js
+{
+// ...
+"Thing":[["addr", "address"]["balance","uint64"]],
+// ...
+}
+```
+
+An example `UserDefinedType` that is a simple type alias:
+
+```py
+from typing import Literal
+from pyteal import abi
+
+HashDigest = abi.StaticBytes[Literal[32]]
+
+```
+
+Produces the following entry in the `TypeSpec`
+```js
+{
+  // ...
+  "HashDigest":"bytes[32]",
+  // ...
+}
+```
+
+
+### Error Specification
+
+TODO: number as key will fail in certain places, lets do something else?
+
+The Error specification contains a mapping of `PC` (or `Program counter` TODO: Link to description) to a human readable string that describes the error condition.  
+
+When a node returns a logic eval error, the `pc` specified in the error message should be looked up in the ErrorSpec map to find the relevant error message.
+
+```ts
+// TODO: more complex spec for errors? what else would we want? 
+type ErrorMsg = string;
+type ErrorSpec = Record<number, ErrorMsg>;
+```
+
+To build an error spec, compile the source teal program with its source map, and map the `PC` of any `assert` or `err` opcodes with a comment on the immediately preceding line into the `ErrorSpec` map. 
+
+ex: the following will produce an entry in the error spec like `{..., 10: "// Sorry but 2 is not less than 1", ...}` assuming the `PC` of the `assert` op is 10.
+```
+int 1    
+int 2
+<
+// Sorry but 2 is not less than 1
+assert
+```
```

Comment:
Just chiming in to say that I really like the idea of having a structured way of encoding/enumerating App’s errors.

Something like: if I fail on `PC=10` we can use `sourcemap` to easily return something like:

`ERR-010: This is a bad error` 

to the client so that an eventual App front-end could provide better error handling experience.

Also this would enable more "granularity" at testing level, to prove that a given transaction is failing due to `ERR-010`, as expected, and not due to `ERR-042` or anything else.

Reply:
Good point, ill just caution against adding the pc to the error message if that is what the 10 in  `ERR-010`  is meant to signify since the pc could change with every edit to the program

Reply:
Right, also maybe want to have the same error code at different PCs, like `ERR-X: Insufficient funds` that can be used in the execution path of different methods.

Reply:
The important thing is having a way to enumerate the error causes. At PyTeal (or Beaker) / SDK level some helper classes could facilitate errors mapping, something like:

```python
Assert(Int(2) < Int(1), comment=Errors.MATH)
```

The `Errors` class (or something similar) could be used to generate the related Errors AppSpec.

Reply:
This is indeed what I did for the AMM example:
https://github.com/algorand-devrel/beaker/blob/master/examples/amm/amm.py#L47-L57